src/util/base.cc

```diff
@@ -0,0 +1,231 @@
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY        //
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF    //
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL     //
+ * THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,       //
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,               //
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS    //
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,          //
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF    //
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.               //
+ */
+
+#include "util/base.h"
+#include <assert.h>
+#include "core/util/base64.h"
+
+namespace bpo = boost::program_options;
+
+#define BASE32_BUF_SIZE 400
+#define BASE64_BUF_SIZE 120
+
+static_assert(BASE32_BUF_SIZE % 40 == 0, "Invalid BASE32_BUF_SIZE");
+static_assert(BASE64_BUF_SIZE % 12 == 0, "Invalid BASE64_BUF_SIZE");
+
+BaseCommand::BaseCommand() : m_desc("General options")
+{
+  m_desc.add_options()("help,h", "produce this help message")(
+      "type,t", bpo::value<std::string>(&m_opt_type), "encode/decode")(
+      "infile,i", bpo::value<std::string>(&m_opt_infile), "input file")(
+      "outfile,o", bpo::value<std::string>(&m_opt_outfile), "output file");
+}
+
+void BaseCommand::PrintUsage(const std::string& name) const
+{
+  std::cout << "Syntax: " << name << " <options> encode <inFile> <outfile>"
+            << std::endl;
+  std::cout << "or    : " << name << " <options> decode <inFile> <outfile>"
+            << std::endl;
+  std::cout << std::endl;
+  std::cout << m_desc << std::endl;
+}
+
+template <typename ToBase, typename ToByte, typename SizeConst>
+bool process(
+    ToBase ByteStreamToBase,
+    ToByte BaseToByteStream,
+    SizeConst Size,
+    bool encode,
+    std::istream& input,
+    std::ostream& output)
+{
+  std::uint8_t in_buffer[Size + 1];
+  std::uint8_t out_buffer[Size * 2 + 1];
+  size_t ret(0);
+  memset(in_buffer, 0, sizeof(in_buffer));
+  memset(out_buffer, 0, sizeof(out_buffer));
+  do
+    {
+      input.read(reinterpret_cast<char*>(in_buffer), Size);
+      if (input.bad())
+        {
+          std::cerr << "Error while reading input ! " << strerror(errno)
+                    << std::endl;
+          return false;
+        }
+
+      if (encode)
+        ret = ByteStreamToBase(
+            in_buffer,
+            input.gcount(),
+            reinterpret_cast<char*>(out_buffer),
+            Size * 2);
+      else
+        ret = BaseToByteStream(
+            reinterpret_cast<char*>(in_buffer),
+            input.gcount(),
+            out_buffer,
+            Size * 2);
+
```

Comment:
Maybe check for valid return (was stream successfully processed) and report to user?